### DISCUSSION TITLE
Using SymPy For Website

### DISCUSSION THREAD

**Discussion by Ephraim-Bryski** [sympy/sympy#21225] [Q&A]
I'm interested in using a CAS for a website. While there are JavaScript CAS, SymPy has generally been recommended to me. Would it be possible to use SymPy and pyjs in combination for a website? Is this often done, and are there any issues associated with it?

**Comment by oscarbenjamin** [comment]
This seems like a more suitable question for the mailing list.

**Comment by asmeurer** [comment]
I don't know if this has been tried before. I would expect something like this would not be easy, but please report back if you have any success with it.

**Reply by Ephraim-Bryski** [reply]
Thank you! I don't think I would be one to take on that challenge though (I really just know basic coding, so I don't think I would know how to approach that sort of problem). I know this is no longer directly related to the original question, but do you libraries which you generally recommend to use on a browser?

**Reply by asmeurer** [reply]
I don't know of the best way to run a CAS in-browser. I know some people have experimented with SymPy and Pyodide, but I don't know the status of it. There may be Javascript libraries, but likely much simpler in terms of functionality compared to SymPy. 

My suggestion would be to run SymPy on a server backend.

**Reply by Ephraim-Bryski** [reply]
Thank you once again. I made a website already ([seeciv.com](seeciv.com)), but so far it's only front-end; I have no experience coding back-end. I'm using Netlify to host the website.

Would it be difficult to develop the back-end? Would I be coding entirely in Python (which I am familiar with) or other languages as well like PHP (which I'm not)? Are there good resources for this (I've found general [articles](https://www.educative.io/blog/web-development-in-python) on backend with Python, but I'm not sure if they're specific to what I want to do)?

I also plan on using a WYSISWYG equation editor in JavaScript like MathQuill in combination with the CAS. Would it be simple to communicate between the CAS and equation editor if they're written in Python and Javascript respectively?

**Reply by asmeurer** [reply]
Unfortunately, these are not really things I'm an expert at, and most other members of the SymPy community won't be either. You may need to take your questions to other forums to get good answers. My suggestion would be to find open source projects that are doing something similar to what you want to do and see how they do it.

**Reply by oscarbenjamin** [reply]
If you want to reach people who might have implemented something like this already then I think the mailing list is probably a better place than github.

**Reply by stefnotch** [reply]
@Ephraim-Bryski I posted an answer just below, hopefully you find it interesting. If you've got any questions or want to re-use any of the code, by all means, do ping me.

**Comment by asmeurer** [comment]
> This seems like a more suitable question for the mailing list.

We can convert issues like this into discussions.

**Comment by stefnotch** [comment]
This is actually doable, *however* you'll want to use [pyodide](https://github.com/pyodide/pyodide). Pyodide, while a bit on the [heavy](https://github.com/pyodide/pyodide/issues/646) side, does support just about any Python package and ships with Sympy by default.
You can go to the [REPL](https://pyodide.org/en/latest/console.html) and try something like 
```py
from sympy import *
x = symbols('x')
factor(-2*x**2 + x + 21)
```

<hr>

*Blatant self advertisement:* There is a project that does just that, it's called [QuantumSheet and works surprisingly well](https://stefnotch.github.io/quantum-sheet/?document-url=https://gist.githubusercontent.com/stefnotch/fe8a80a57d2ee9211b7a24ab7fcc9ce2/raw/document.qs).
In case you're interested, it basically takes [mathlive for user input](https://github.com/arnog/mathlive), the very experimental [cortexjs compute-engine to parse it](https://cortexjs.io/compute-engine/demo/) and finally generates sympy code.

So, the user types the left-hand side of
![image](https://user-images.githubusercontent.com/10220080/130362614-9133a124-28c5-4c42-b4b0-a03cb277aa8f.png)

Then it generates something along the lines of
```py
sympy.factor(
	sympy.Add(sympy.Mul(sympy.Integer(-2),sympy.Pow(_x,sympy.Integer(2))),_x,sympy.Integer(21))
		.subs({})
)
```
And that gets sent to Pyodide, which is running in a web worker. As the final step, the Sympy result is returned, converted and displayed.

**Comment by boniSantana** [comment]
With Pyodide and Observable you can put sympy and show results easy for a educational/experimental level, I use this alternative a lot.

Examples:

https://observablehq.com/@fedesantana/sintesis-de-funciones-transferencia
https://observablehq.com/@fedesantana/graficos-interactivos-de-tc2/2?collection=@fedesantana/tc2
https://observablehq.com/@fedesantana/ts11-sintesis-de-funciones-transferencias?collection=@fedesantana/tc2

**Comment by Ephraim-Bryski** [comment]
I came across this old discussion and I thought I might as well say what ended up happening.

I asked a question for a site to solve systems of equations and show the substitution steps, which you can now see at equationvisuals.com. The full codebase is here: [equation-visuals](https://github.com/Ephraim-Bryski/equation-visuals/tree/main).

I initially tried using a CAS, sympy and then nerdamer. I had a lot of trouble coercing it to manipulate the equations as I needed so I ended up writing my own very scrappy computer algebra system. You can see cas.js and solver.js for the computer algebra system and system of equation solver, respectively.

In retrospect, it was probably not well designed, but anyone is free to look at or take any of the code and use it however they like if it's at all helpful.
